src/goto-instrument/contracts/assigns.cpp

```diff
@@ -112,13 +111,30 @@ void assigns_clauset::add_target(const exprt &target_expr)
 
 void assigns_clauset::remove_target(const exprt &target_expr)
 {
-  targets.erase(targett(*this, targett::normalize(target_expr)));
+  write_set.erase(targett(*this, target_expr));
+}
+
+void assigns_clauset::add_freely_assignable_expr(const exprt &expr)
+{
+  auto insertion_succeeded = freely_assignable_set.emplace(*this, expr).second;
+
+  if(!insertion_succeeded)
+  {
+    log.warning() << "Ignored duplicate expression '"
+                  << from_expr(ns, expr.id(), expr) << "' in assigns clause at "
+                  << expr.source_location().as_string() << messaget::eom;
+  }
+}
+
+void assigns_clauset::remove_freely_assignable_expr(const exprt &expr)
+{
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
```

Comment:
```suggestion
  freely_assignable_set.erase(targett(*this, expr));
```
since `normalize` is called by the constructor.

Reply:
Done.